include/secp256k1.h

```diff
@@ -131,6 +131,30 @@ typedef int (*secp256k1_nonce_function_t)(
 #  define SECP256K1_ARG_NONNULL(_x)
 # endif
 
+# if defined _MSC_VER || defined __CYGWIN__
+#  define SECP256K1_HELPER_LOCAL
+#  define SECP256K1_HELPER_DLL_IMPORT __declspec(dllimport)
+#  define SECP256K1_HELPER_DLL_EXPORT __declspec(dllexport)
+# else
+#  if __GNUC__ >= 4
+#   define SECP256K1_HELPER_LOCAL      __attribute__ ((visibility ("hidden")))
+#   define SECP256K1_HELPER_DLL_IMPORT __attribute__ ((visibility ("default")))
+#   define SECP256K1_HELPER_DLL_EXPORT __attribute__ ((visibility ("default")))
+#  else
+#   define SECP256K1_HELPER_LOCAL
+#   define SECP256K1_HELPER_DLL_IMPORT
+#   define SECP256K1_HELPER_DLL_EXPORT
+#  endif
+# endif
+
+# if defined SECP256K1_STATIC
+#  define SECP256K1_API SECP256K1_HELPER_LOCAL
+# elif defined SECP256K1_EXPORTS
+#  define SECP256K1_API SECP256K1_HELPER_DLL_EXPORT
+# else
+#  define SECP256K1_API SECP256K1_HELPER_DLL_IMPORT
+# endif
+
```

Comment:
The above matches closely the [GCC example](https://gcc.gnu.org/wiki/Visibility) (bottom of page). There is deviation in the public symbols `SECP256K1_STATIC` and `SECP256K1_EXPORTS`. The latter is modified for better CMake integration (see [comment](https://github.com/bitcoin/secp256k1/pull/223#discussion_r39398451) by @chfast). WRT the former I consider "static" more explicit than "local" and therefore better for publication.